src/runtime/middleware/auth.ts

```diff
@@ -1,32 +1,68 @@
+    if (!matchedRoute) {
+      // Hands control back to `vue-router`, which will direct to the `404` page
+      return
+    }
+  }
+
   /**
-   * Nuxt 3 default behavior is to continue with the route navigation if the middleware does not return a `navigateTo` (see https://nuxt.com/docs/guide/directory-structure/middleware#format).
+   * We cannot directly call and/or return `signIn` here as:
+   * - `signIn` uses async composables under the hood, leading to "nuxt instance undefined errors", see https://github.com/nuxt/framework/issues/5740#issuecomment-1229197529
+   * - if something different than `navigateTo` is returned the navigation will not be blocked, so a content-flash would occur, see https://nuxt.com/docs/guide/directory-structure/middleware#format
+   *
+   * Additionally on the client-side, returning `navigateTo(signInUrl)` leads to a `404` error as the next-auth-signin-page was not registered with the vue-router that is used for routing under the hood. For this reason we need to
+   * manually set `window.location.href` on the client **and then fake return a navigateTo to block navigation (although it will not actually be called, but nuxt magically registers the `navigateTo` return and blocks navigation, avoiding
+   * content-flashes of the protected page)**.
    *
-   * To avoid content-flashes or brief visibility of protected pages, we thus need to:
-   * 1. `await signIn` so that it will block and trigger it's internal `navigateTo` call. Using `return signIn(...)` does not work eventhough it internally returns `navigateTo`!
-   * 2. fake-return a `navigateTo` here to tell nuxt that the current navigation will be replaced with a new navigation, this avoids the content-flash of the projected page (and should in theory never be triggered due to `signIn` navigating away first!)
+   * Additionally on the server-side, we cannot use `navigateTo(signInUrl)` as this uses `vue-router` internally which does not know the "external" sign-in page of next-auth and thus will log a warning which we want to avoid.
    *
-   * See https://github.com/sidebase/nuxt-auth/issues/100 for a discussion on this topic.
-   * */
-  await signIn(undefined, { callbackUrl: to.path, redirect: false }, { error: 'SessionRequired' })
-  return navigateTo(joinPathToApiURL('signin'))
+   *  */
+  const signInUrl = joinURL(authConfig.basePath, 'signin')
+  const url = withQuery(signInUrl, {
+    callbackUrl: to.path,
+    error: 'SessionRequired'
+  })
+
+  // adapted from: https://github.com/nuxt/framework/blob/ab2456c295fc8c7609a7ef7ca1e47def5d087e87/packages/nuxt/src/app/composables/router.ts#L97-L115
+  if (process.server) {
+    const nuxtApp = useNuxtApp()
+    if (nuxtApp.ssrContext && nuxtApp.ssrContext.event) {
+      return nuxtApp.callHook('app:redirected').then(() => sendRedirect(nuxtApp.ssrContext!.event, url, 302))
+    }
+  }
+
+  window.location.href = url
+  return navigateTo(undefined)
```

Comment:
In my tests I noticed that `navigateTo(undefined)` will redirect the user to `/` before redirecting to signIn, in client side. `navigateTo` defaults to `/` when the path is not provided. See https://github.com/nuxt/framework/blob/ab2456c295fc8c7609a7ef7ca1e47def5d087e87/packages/nuxt/src/app/composables/router.ts#L77-L79

To reproduce: go to [page that is always unprotected](http://localhost:3000/always-unprotected), then click  [globally protected page](http://localhost:3000/protected/globally). You will see a content-flash of the home page, and if you go back it will go to home page.

I'm not sure how to solve this :/

Reply:
Looked into it -> figured out that there is a bit less magic than I though: As long as we return a `Promise` that does not resolve immeadiatly, the content of the protected page will not be loaded.

I now set it to 100ms: It should never actually be awaited but will block the navigation for up to `100ms` or until the navigation to the singin-page completes (which should always be faster than 100ms, I hope?).

- Does this resolve the issue?
- Do you think 100ms is enough or should we increase it?

Reply:
Yeah, this actually works!!! But I think 100ms is not enough for slow connections. If I set network to slow 3g in the browser, I can see the content-flash for a long time.

If we set the `setTimeout` to `Infinity`, it solves completely. Can we possible cause a side-effect by using `Infinity` here? I have not used it before haha

Reply:
Woop wooop 🎊 

Good catch - I wanted to avoid using `Infinity`, so as a (good I think?) compromise I adapted the behavior:
- wait for up to 60 seconds
- if that does not work: delegate control back to `vue-router` -> this will still avoid content-flashes. The only downside is that it will log a `404` error in the console because vue-router does not know about the signin page

Reply:
Yeah, I think this is good. Tested here and seems to work fine!